### ISSUE TITLE
RyuJIT/x64  failure

### ISSUE THREAD

**Issue by dsyme** [dotnet/coreclr#12480]
This issue with the JIT64 was reported in the Visual F# repo https://github.com/Microsoft/visualfsharp/issues/2224 a while back - it is possible it has already been fixed though it continues to fail on my machine:

Original code: [Program.txt](https://github.com/Microsoft/visualfsharp/files/697635/Program.txt)

#### Repro steps (from binary zip) ####

Attached a built .exe and fsharp.core.dll :

[repro.zip](https://github.com/Microsoft/visualfsharp/files/697685/repro.zip)

Plus you need this repro.exe.config:
```
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <runtime>
    <legacyUnhandledExceptionPolicy enabled="true" />
    <gcServer enabled="true"/>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="FSharp.Core" publicKeyToken="b03f5f7f11d50a3a" culture="neutral"/>
        <bindingRedirect oldVersion="2.0.0.0-4.4.1.0" newVersion="4.4.1.0"/>
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>
```

#### Repro steps (from source) ####

1. Use the F# complier  to compile the file attached "C:\Program Files (x86)\Microsoft SDKs\F#\4.0\Framework\v4.0\fsc.exe" -o:repro.exe  Program.fs --platform:x64 

2. When we run the exe that is generated from the above command it fails with following error

Unhandled Exception: System.TypeInitializationException: The type initializer for '<StartupCode$repro>.$Program' threw an exception. ---> System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at Program.parseFixTo_BigStruct(FieldsWrapper fields)
   at <StartupCode$repro>.$Program..cctor()
   --- End of inner exception stack trace ---
   at Program.main(String[] argv)

#### Expected behavior ####

Exe should run fine without error

#### Actual behavior ####

Unhandled Exception: System.TypeInitializationException: The type initializer for '<StartupCode$repro>.$Program' threw an exception. ---> System.InvalidProgramException: Common Language Runtime detected an invalid program

#### Known workarounds #### 

If you set the  platform to X86 then there is no exception as below.

"C:\Program Files (x86)\Microsoft SDKs\F#\4.0\Framework\v4.0\fsc.exe" -o:repro.exe  Program.fs --platform:x86 

#### Related information 

Provide any related information 

* Operating system: Windows 10 with anniversary update
* .NET Runtime, 4.6.2
* Editing Tools (e.g. Visual Studio Version) - VS2015 Update 3 

category:correctness
theme:large-methods
skill-level:expert
cost:large

**Comment by RussKeldorph** [comment]
@russellhadley Have you looked into this or the original issue (Microsoft/visualfsharp#2224)?

@dsyme If I'm reading the discussion in Microsoft/visualfsharp#2224 correctly, the title of this issue should be "RyuJIT/x64 ..." as that is what you should get when `useLegacyJit` is disabled on Win10.  JIT64 is the name of the JIT you get when `useLegacyJit` is enabled.  FYI.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
I'll take a look.

**Comment by AndyAyersMS** [comment]
Repros on desktop with 4.7.2101.1. Succeeds with legacy jit.

When jitting `Program:parseFixTo_BigStruct(ref):ref` RyuJit hits this internal limitation:
```
IMPL_LIMITATION("JIT doesn't support offsets larger than 255 into valuetypes for local vars > 32767\n");
```
with varNum 32857 and offset 256.

The program keeps a lot of values on the evaluation stack across branches, at one point the jit needs 332 temps, and indeed maxstack for the method is 338. So likely the underlying issue is simply the large number of values live on the stack.
```
...
lvaGrabTemps(329) returning 30191..30519 (long lifetime temps) called for IL Stack Entries
lvaGrabTemps(329) returning 30520..30848 (long lifetime temps) called for IL Stack Entries
lvaGrabTemps(330) returning 30849..31178 (long lifetime temps) called for IL Stack Entries
lvaGrabTemps(330) returning 31179..31508 (long lifetime temps) called for IL Stack Entries
lvaGrabTemp returning 31509 (V31509 tmp31508) called for NewObj constructor temp.
lvaGrabTemps(331) returning 31510..31840 (long lifetime temps) called for IL Stack Entries
lvaGrabTemps(331) returning 31841..32171 (long lifetime temps) called for IL Stack Entries
lvaGrabTemp returning 32172 (V32172 tmp32171) called for NewObj constructor temp.
lvaGrabTemps(332) returning 32173..32504 (long lifetime temps) called for IL Stack Entries
```
Same thing happens with minopts so the issue is not caused by optimization.

The jit allocates a new set of locals for each control flow join where there are values live in on the evaluation stack. Having stack depth in the 100s across 100s of branches quickly leads to 10,000s of temps and banging up against internal limits in the jit.

@dsyme one possible fix is to change your IL generation strategy to not keep so many values on the evaluation stack across control flow -- instead save values to locals and reload where needed.  Not sure if this is viable/feasible but it would avoid hitting limits like the above.

The jit might also be able to re-use stack temps to achieve a similar local compression. Not sure how viable/feasible this is either. For instance, if at some join all predecessors are going to copy from the same prior temp into a new temp, the jit could just re-use the old temp. This would help in cases like the one here where most blocks mainly pass through existing stacks, and would also compress the IR quite a bit as there would be many fewer many spill/reload statements.

For example, given a "diamond" control flow chain A -> (B,C) -> D -> (E,F) -> G, where A and D don't touch the stack, B,C,E and F all push one item, today we'd generate code like this:
```
B: store 2 items to locals t0, t1, branch to D
C: store 2 stack items to locals t0, t1, branch to D
D: load t0, t1 onto stack .... 
E: store 3 items to locals t2, t3, t4, branch to G
F: store 3 items to locals  t2, t3, t4, branch to G
G: load t2, t3, t4 onto stack
```
and this could be 
```
B: store 2 items to locals t0, t1, branch to D
C: store 2 stack items to locals t0, t1, branch to D
D: load t0, t1 onto stack .... 
E: store 3 items to locals t0, t1, t2 branch to G   // reuse t0, t1 
F: store 3 items to locals  t0, t1, t2, branch to G  // reuse t0, t1
G: load t0, t1, t2 onto stack
```
@dotnet/jit-contrib any other ideas?

**Comment by JosephTremoulet** [comment]
> any other ideas?

Maybe when the number of stack slots coming into a join is above some threshold, we could dummy up a struct and pass the slots as fields of a single struct temp, rather than individual temps?  So that in the 100s of live values across 100s of blocks case, rather than 10,000 temps we'd have 100s of temps that are structs with 100s of fields.  It would keep our lclVar table smaller, and possibly pessimize but speed up things like liveness analysis that would model the structs coarsely rather than field-by-field.

Though perhaps that would be difficult to implement, since we wouldn't have a `clsHnd` to describe the dummied-up struct type...

**Comment by AndyAyersMS** [comment]
My possibly-fix-in-the-jit example is a bit off, though I don't think it alters anything fundamental. There is some attempt in the importer to reuse temps across "spill cliques" that I need to understand better...

**Comment by AndyAyersMS** [comment]
Relevant C# example:
```C#
using System;
using System.Runtime.CompilerServices;

class S
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static int F(int x, int y, int z)
    {
        return x + y + z;
    }

    public static int Main(string[] args)
    {
        int L = args.Length;
        int result = F( (L > 0 ? 1 : 3), (L > 1 ? 2 : 4), (L > 2 ? 11 : 13));
        return result;
    }
}
```
with disassembly below. Note the large number of temps and the register shuffling it facilitates.
```asm
; Assembly listing for method S:Main(ref):int
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T00] (  3,  3   )     ref  ->  rcx         class-hnd
;  V01 loc0         [V01,T01] (  4,  4   )     int  ->  rcx
;  V02 tmp0         [V02,T02] (  3,  2   )     int  ->  rdx
;  V03 tmp1         [V03,T03] (  3,  2   )     int  ->  rdx
;  V04 tmp2         [V04,T04] (  3,  2   )     int  ->  rdx
;  V05 tmp3         [V05,T05] (  3,  2   )     int  ->   r8
;  V06 tmp4         [V06,T06] (  3,  2   )     int  ->  rdx
;  V07 tmp5         [V07,T07] (  3,  2   )     int  ->   r8
;  V08 tmp6         [V08,T08] (  3,  2   )     int  ->  rcx
;  V09 tmp7         [V09,T09] (  3,  2   )     int  ->   r8
;  V10 tmp8         [V10,T10] (  3,  2   )     int  ->  rax
;  V11 OutArgs      [V11    ] (  1,  1   )  lclBlk (32) [rsp+0x00]
;
; Lcl frame size = 40

G_M16040_IG01:
       4883EC28             sub      rsp, 40

G_M16040_IG02:
       8B4908               mov      ecx, dword ptr [rcx+8]
       85C9                 test     ecx, ecx
       7F07                 jg       SHORT G_M16040_IG03
       BA03000000           mov      edx, 3
       EB05                 jmp      SHORT G_M16040_IG04

G_M16040_IG03:
       BA01000000           mov      edx, 1

G_M16040_IG04:
       83F901               cmp      ecx, 1
       7F08                 jg       SHORT G_M16040_IG05
       41B804000000         mov      r8d, 4
       EB06                 jmp      SHORT G_M16040_IG06

G_M16040_IG05:
       41B802000000         mov      r8d, 2

G_M16040_IG06:
       83F902               cmp      ecx, 2
       7F09                 jg       SHORT G_M16040_IG07
       8BCA                 mov      ecx, edx
       B80D000000           mov      eax, 13
       EB07                 jmp      SHORT G_M16040_IG08

G_M16040_IG07:
       8BCA                 mov      ecx, edx
       B80B000000           mov      eax, 11

G_M16040_IG08:
       418BD0               mov      edx, r8d
       448BC0               mov      r8d, eax
       E8AEFBFFFF           call     S:F(int,int,int):int
       90                   nop

G_M16040_IG09:
       4883C428             add      rsp, 40
       C3                   ret

; Total bytes of code 80, prolog size 4 for method S:Main(ref):int
```

**Comment by AndyAyersMS** [comment]
Did a bit of instrumentation and prospecting on how often this is and how many temps might be saved. Don't have the full data yet, but when crossgenning corelib, the jit allocates 10849 stack merge temps and my analysis says it only needed to allocate 6217.

So far all the examples where we see nontrivial depths at merges are similar to the above.

Am going to figure out what fraction of all (importer-allocated) temps these stack merge temps represent and what fraction of initial IR is trafficking in unneeded temps (each unneeded merge temp will give rise to slightly over two assignment statements, on average).

Assuming that cleaning this up would reduces importer IR volume by a percent or more it is probably worth addressing it in the jit. The fix may not be easy but I have some notion that we can track the spill temp IDs for each stack slot (which will no longer be contiguous) in the entry stack states.

**Comment by tamlin-mike** [comment]
@AndyAyersMS For your example from Aug 3, instead of all that jumping (2 jmp's for each branch), what about instead using cmp/mov reg, .../jX/add y, when both terms are (integral-only?) constants?

While I haven't profiled it, to see if add to a reg assigned-to two instructions earlier would introduce a pipeline stall, at first blush it seems it could possibly be faster (in addition to smaller - which is obviously a good thing).

**Comment by AndyAyersMS** [comment]
@tamlin-mike agree, it would be better. There are lots of clever ways to manifest literal integers, especially if you have similar integers "nearby" already or can create some values cheaply. Unfortunately the jit doesn't do this kind of thing.

We have debated the idea of creating a post-lower machine-dependent "peephole" optimizer that would provide a natural home for this kind of optimization, but it hasn't gotten past the idea stage.

**Comment by tamlin-mike** [comment]
I think such an optimizer, or simply architecture, would be beneficial. Maybe especially for hot-spots (provided there is a hot-spot finder). 

Doing it that way (loosely decoupled from main JIT code) could also
* enable way easier testing, of
  * correctness
  * performance of generated code
  * performance of generator/optimizer
*  provide a playground to actually try out ideas without affecting the release branch
   * this could also be a nice place to document what has been tested/tried out, and what the result was.
* possibly make it easier for external contributors/contributions.

Just some brainstorming thoughts.

**Comment by mikedn** [comment]
`L > 0 ? 1 : 3` is just
```asm
test    edi, edi
setle   al
lea     eax, [rax + rax + 1]
```
That's something that if-conversion could handle. Ironically, in my experiment if-conversion was sometimes blocked by this particular importing issue. I worked around it, at least partially, by detecting identical lclvar assignments and hoisting them out of the blocks subjected to if-conversion.

**Comment by AndyAyersMS** [comment]
@dsyme can you weigh in on how impactful this is for F#? 

I don't see any easy fixes in the jit right now. So it is unlikely to get fixed in Core 2.1.

**Comment by AndyAyersMS** [comment]
Pinging some other F# folks -- @KevinRansom?

**Comment by dsyme** [comment]
@AndyAyersMS We've only had the one user report of this error.  However I would hope to see it fixed in some way, or some specific size limitation documented.

Would it be possible to at least give a  better error message in the exception, e.g. giving the name of the method and some hint to the user about what size limitation has been hit? i.e. then the user can reduce the size of initialization code or the size of a method?  

thanks!
don

p.s. sorry not to see the @dsyme shoutout - please feel free to email

**Comment by AndyAyersMS** [comment]
The jit struggles with IL when there is a lot of state on the stack across non-trivial control flow. 

It manages the cross-block state by generating what amounts to a series of stloc/ldloc operations at the respective block boundaries. This generation happens early in the jit before it can do any real analysis. So the conversion is inefficient and can at times produce a lot of locals (~30,000 in this particular case). And in extreme cases the sheer volume of these locals runs into a representation limit in the jit's IR, which is why the jit eventually fails.

So as a fix, it would be good to limit how much state gets pended to the IL stack by FSC. A maxstack of 338 is an indication of potential trouble. I would expect that FSC should be able to spill and restore these sorts of stacks with a fairly small number of locals.

I'll look into what it will take for the runtime to convey the name of the method being jitted when jitting fails -- that seems like a potentially useful fact.

I don't know how successful we can be at describing why the jit has failed, at least in terms that are actionable for end users. For instance we'd probably describe this jit failure as "too many local variables" but the user method has no local variables.

**Comment by AndyAyersMS** [comment]
FYI this repros in Core with `dotnet new console -lang F#` too.

The stack trace associated with the error message identifies the problematic method:
```
Unhandled Exception: System.TypeInitializationException: The type initializer for '<StartupCode$zz>.$Program' threw an exception. ---> System.InvalidProgramException: Common Language Runtime detected an invalid program.
   at Program.parseFixTo_BigStruct(FieldsWrapper fields)
   at <StartupCode$zz>.$Program..cctor() in D:\bugs\12480\zz\Program.fs:line 375
   --- End of inner exception stack trace ---
   at Program.main(String[] argv) in D:\bugs\12480\zz\Program.fs:line 379
```

**Comment by dsyme** [comment]
Cc @cartermp. The user will work out that something is too big in that method. At least the will know its a size thing..

**Comment by AndyAyersMS** [comment]
@jkotas any reason why we can't unblock x64 in `void ThrowExceptionForJit(HRESULT res)` so that jit internal errors get reported as such?

**Comment by AndyAyersMS** [comment]
Seems to work well enough:
```
Unhandled Exception: System.InvalidProgramException: JIT Compiler encountered an internal limitation.
   at InvalidProgramExceptionReproduction.Program.Map_0(IEnumerable`1 docs)+MoveNext()
   at InvalidProgramExceptionReproduction.Program.Main(String[] args)
```

**Comment by jkotas** [comment]
I do not see a problem with it. It looks like JIT64 compat quirk.

**Comment by AndyAyersMS** [comment]
We now will produce a slightly better error message when we hit jit internal limitations (#16320). So I'm going to move this issue out to future.

**Comment by AndyAyersMS** [comment]
Brian's fixes in #17793 (for master/3.0) and #21493 (for upcoming 2.2 servicing) allows the jit to finish compiling this case.

So I am going to close this.
